extensions/amp-cxense/0.1/amp-cxense.js

```diff
@@ -0,0 +1,276 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {CSS} from '../../../build/amp-cxense-0.1.css';
+import {getLengthNumeral, isLayoutSizeDefined} from '../../../src/layout';
+import {loadPromise} from '../../../src/event-helper';
+import {setStyles} from '../../../src/style';
+import {timer} from '../../../src/timer';
+
+const DEFAULT_AMD_SRC = location.protocol + '//embed.widget.cx/amd.js';
+const DEFAULT_EMBED_APP = '/app/player/m4/dist/';
+const PRE_CONNECT_URLS = [DEFAULT_AMD_SRC].concat([
+    location.protocol + '//api.widget.cx',
+    location.protocol + '//api.ramp.com' // todo: rm when safe
+]);
+const DATA_ATTR_NAME_MIDFIX = '-ramp-'; // todo: change to -cx- when safe
+
+let skipTransfer = {
+    'data-isplayer': true,
+    'data-noui': true,
+    'data-placeholder': true,
+    'class': true,
+    'layout': true
+};
+
+let _getLongDataAttributeName = (name) => {
+    return name.replace(/^data-/i, 'data' + DATA_ATTR_NAME_MIDFIX);
+};
+
+class AmpCxense extends AMP.BaseElement {
+
+    /** @override */
+    preconnectCallback(onLayout) {
+        let self = this;
+        PRE_CONNECT_URLS.forEach((url) => {
+            self.preconnect.url(url, onLayout);
+        });
+    }
+
+    /** @override */
+    isLayoutSupported(layout) {
+        return isLayoutSizeDefined(layout);
+    }
+
+    /** @override */
+    buildCallback() {
+        this.element.className += ' amp-cxense';
+
+        let src = this.element.getAttribute('src');
+
+        let _widgetId = this._getDataAttribute('widget-id');
+        let _embed = this._getDataAttribute('embed');
+        let _module = this._getDataAttribute('module');
+
+        if (!src && !_widgetId && !_embed && !_module) {
+            // if none of these is specified, the default behavior is to load the app widget.
+            this.element.setAttribute('data-embed', DEFAULT_EMBED_APP);
+            this.element.setAttribute('data-isplayer', true);
+        }
+
+        /** @private @const {string} */
+        this._src =  src || DEFAULT_AMD_SRC;
+        // necessary, otherwise it would confuse AMP
+        this.element.removeAttribute('src');
+
+        /** @private @const {string} */
+        this._placeholder = this.element.getAttribute('data-placeholder') !== "false";
```

Comment:
some of our widgets should not be visible initially, this gives the option to hide it and show it later while waiting for something, for example, a page toaster that shows up when scroll at the bottom of the page. This one does not require a placeholder.